src/script/sign.h

```diff
@@ -68,6 +71,434 @@ struct SignatureData {
+                        }
+                        // Retrieve sha256 from key
+                        uint256 hash(std::vector<unsigned char>(key.begin() + 1, key.end()));
+
+                        // Read in the redeemscript
+                        std::vector<unsigned char> witnessscript_bytes;
+                        witnessscript_bytes.resize(value_len);
+                        Span<unsigned char> witnessscript_data = MakeSpan(witnessscript_bytes);
+                        s >> witnessscript_data;
+                        CScript witnessscript(witnessscript_bytes.begin(), witnessscript_bytes.end());
+
+                        // Check that the witnessscript sha256 matches the one provided
+                        sha256_hasher.Write(&witnessscript_bytes[0], witnessscript_bytes.size());
+                        unsigned char ws_sha256_data[sha256_hasher.OUTPUT_SIZE];
+                        sha256_hasher.Finalize(ws_sha256_data);
+                        uint256 ws_sha256(std::vector<unsigned char>(ws_sha256_data, ws_sha256_data + sha256_hasher.OUTPUT_SIZE));
+                        if (hash != ws_sha256) {
+                            throw std::ios_base::failure("Provided sha256 does not match the witnessscript's sha256");
+                        }
+
+                        // Add to map
+                        witness_scripts.emplace(hash, witnessscript);
+                    } else {
+                        // Make sure that the key is the size of pubkey + 1
+                        if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
+                            throw std::ios_base::failure("Size of key was not the expected size for the type partial signature pubkey");
+                        }
+                        // Read in the pubkey from key
+                        CPubKey pubkey(key.begin() + 1, key.end());
+                        if (!pubkey.IsFullyValid()) {
+                           throw std::ios_base::failure("Invalid pubkey");
+                        }
+
+                        // Read in the signature from value
+                        std::vector<unsigned char> sig;
+                        sig.resize(value_len);
+                        Span<unsigned char> sig_data = MakeSpan(sig);
+                        s >> sig_data;
+
+                        // Add to list
+                        input.partial_sigs.emplace(pubkey, sig);
+                    }
+                    break;
+                // BIP 32 HD Keypaths and sighash types
+                case PSBT_BIP32_KEYPATH_SIGHASH:
+                    if (in_globals) {
+                        // Make sure that the key is the size of pubkey + 1
+                         if (key.size() != CPubKey::PUBLIC_KEY_SIZE + 1 && key.size() != CPubKey::COMPRESSED_PUBLIC_KEY_SIZE + 1) {
```

Comment:
In "Implement PSBT structures and un/serliazation methods per BIP 174":

Nit: mis-aligned indentation. (remove 1 space I think?)

Reply:
Done